compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Utils/ChannelBdIdGenerator.h"
+#include "iree-amd-aie/aie_runtime/iree_aie_runtime.h"
+
+#define DEBUG_TYPE "iree-amdaie-assign-npu-dma-bd-ids"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+/// Assign BD ids to NPU dma operations using the BD generator
+LogicalResult assignNpuDmaBdIds(AMDAIE::WorkgroupOp workgroupOp) {
+  IRRewriter rewriter(workgroupOp->getContext());
+
+  // Get the device model.
+  std::optional<AMDAIEDevice> device = getConfigAMDAIEDevice(workgroupOp);
+  if (!device)
+    return workgroupOp->emitOpError()
+           << "could not find an AMDAIEDevice attribute";
+  AMDAIEDeviceModel deviceModel = AMDAIE::getDeviceModel(device.value());
+
+  // Create a BD ID generator for every shim tile.
+  DenseMap<Value, ChannelBdIdGenerator> shimTileToGeneratorMap;
+  workgroupOp->walk([&](AMDAIE::TileOp tileOp) {
+    std::optional<int64_t> col = getConstantIntValue(tileOp.getCol());
+    std::optional<int64_t> row = getConstantIntValue(tileOp.getRow());
+    if (col && row && deviceModel.isShimNOCTile(col.value(), row.value())) {
+      ChannelBdIdGenerator generator(
+          deviceModel.getChannelToValidBdIds(AMDAIETileType::SHIMNOC));
+      shimTileToGeneratorMap[tileOp.getResult()] = std::move(generator);
+    }
+  });
+
+  // Walk `amdaie.npu_dma_cpy_nd` and  `amdaie.dma_wait` operations and assign
+  // and release BD IDs when encountering the respective using the tile BD ID
+  // generators initialized earlier.
+  AMDAIE::ControlCodeOp controlCodeOp = workgroupOp.getControlCode();
+  WalkResult res = controlCodeOp->walk([&](Operation *op) {
+    if (auto npuDmaOp = dyn_cast<AMDAIE::NpuDmaCpyNdOp>(op)) {
+      AMDAIE::CircularDmaCpyNdOp inputDma = npuDmaOp.getDmaCpyNdOp();
+
+      // Utility to retrieve a TileOp from a vector of tile values, while doing
+      // appropriate verification.
+      auto getGeneratorTileOp = [&](const SmallVector<Value> &tiles,
+                                    AMDAIE::TileOp &tileOp) -> LogicalResult {
+        if (tiles.size() != 1) {
+          return npuDmaOp.emitOpError()
+                 << "operating on multiple tiles is not supported";
+        }
```

Comment:
is multiple tiles a planned extension or something?

Reply:
Multiple tiles currently only happens on the AIE core side to represent broadcasting data to multiple cores. It shouldn't come up here (LR) but that's why the logical objectfifo can have multiple tiles assigned and therefore `SmallVector<Value>`.